pkg/agent/externalnode/external_node_controller.go

```diff
@@ -0,0 +1,753 @@
+		uplinkName := genUplinkInterfaceName(ifName)
+		iface, err := c.createOVSPortsAndFlows(uplinkName, ifName, eeNamespace, eeName, ips)
+		if err != nil {
+			return err
+		}
+		c.ifaceStore.AddInterface(iface)
+		return nil
+	}
+	klog.InfoS("Updating OVS port data", "ifName", ifName, "eeName", eeName, "ip", ips)
+	iface, err := c.updateOVSPortsData(hostIface, eeName, ips)
+	if err != nil {
+		return err
+	}
+	c.ifaceStore.AddInterface(iface)
+	return nil
+}
+
+func (c *ExternalNodeController) updateExternalNode(obj interface{}, curEn *v1alpha1.ExternalNode) error {
+	klog.InfoS("Updating ExternalNode", "ExternalNode", klog.KObj(curEn))
+	preEn := obj.(*v1alpha1.ExternalNode)
+	preEEName := externalnode.GenExternalEntityName(preEn)
+	preIfName, err := getHostInterfaceName(preEn.Spec.Interfaces[0])
+	if err != nil {
+		return err
+	}
+	curEEName := externalnode.GenExternalEntityName(curEn)
+	curIfName, err := getHostInterfaceName(curEn.Spec.Interfaces[0])
+	if err != nil {
+		return err
+	}
+	if preIfName != curIfName {
+		klog.InfoS("Found interface name is changed", "preName", preIfName, "curName", curIfName)
+		if err = c.deleteInterface(preIfName); err != nil {
+			return err
+		}
+		if err = c.addInterface(curIfName, curEn.Namespace, curEEName, curEn.Spec.Interfaces[0].IPs); err != nil {
+			return err
+		}
+	} else if (!reflect.DeepEqual(preEn.Spec.Interfaces[0].IPs, curEn.Spec.Interfaces[0].IPs)) || (preEEName != curEEName) {
+		klog.InfoS("Found interface configuration is changed", "preIPs", preEn.Spec.Interfaces[0].IPs, "preEEName", preEEName,
+			"curIPs", curEn.Spec.Interfaces[0].IPs, "curEEName", curEEName)
+		if err = c.addInterface(curIfName, curEn.Namespace, curEEName, curEn.Spec.Interfaces[0].IPs); err != nil {
+			return err
+		}
+	}
+	c.syncedExternalNodes.Add(curEn)
+	// Notify the ExternalNode update event to NetworkPolicyController.
+	c.nodeUpdateNotifier.Notify(curEn)
+	return nil
+}
+
+func (c *ExternalNodeController) deleteExternalNode(key string) error {
+	klog.InfoS("Deleting ExternalNode", "key", key)
+	obj, exists, _ := c.syncedExternalNodes.GetByKey(key)
+	if !exists {
+		klog.InfoS("Skipping ExternalNode deletion as it hasn't been synced", "ExternalEntityKey", key)
+		return nil
+	}
+	en := obj.(*v1alpha1.ExternalNode)
+	ifName, err := getHostInterfaceName(en.Spec.Interfaces[0])
+	if err != nil {
+		return err
+	}
+	if err := c.deleteInterface(ifName); err != nil {
+		return err
+	}
+	c.syncedExternalNodes.Delete(en)
+	return nil
+}
+
+func (c *ExternalNodeController) deleteInterface(ifName string) error {
+	hostIface, portExists := c.ifaceStore.GetInterfaceByName(ifName)
+	if !portExists {
+		klog.InfoS("Skipping deleting host interface since it doesn't exist ", "ifName", ifName)
+		return nil
+	}
+	klog.InfoS("Deleting interface", "ifName", ifName)
+	if err := c.removeOVSPortsAndFlows(hostIface); err != nil {
```

Comment:
As part of deleting externalnde, agent is restoring the interface configuration back to Host VM. 
Should we stop/restart antrea-agent for both Linux and Windows, to keep the end-end workflow same?

Reply:
But how you plan to restart agent? Should that be a manual step by users, or you mean the agent should exit itself? I feel the latter is not a good way.

Reply:
We provide a sample to register antrea-agent as a system service on both Linux and Windows. If agent process is quitted but the service itself is not stopped, the antrea-agent process woule be restarted automatically by the service.

But for the scenario that no system service is registered, user should restart the process by himself.

Reply:
I just feel it is strange the agent process exists at config change, and relying on another solution to restart it. Probably we should do that on Linux. If we need to do that on Windows, please add clear log message before exiting.

Reply:
@tnqn @jianjuns I would like to confirm that if you think we should also stop agent for Linux to be consistent with Windows? For now, I do not let agent restart for Linux. I would change this part if you think the behavior should be consistent for these two operating systems.

Reply:
Have confirmed with Jianjun and Quan and we prefer to keep existing behavior and do not let agent exit for Linux.